src/db/ai.rs

```diff
@@ -60,40 +83,107 @@ pub fn create_or_increment_limit(
-            insert_into(schema::ai_help_limits::table)
+            insert_into(limits::table)
                 .values(&limit)
-                .on_conflict(schema::ai_help_limits::user_id)
+                .on_conflict(limits::user_id)
                 .do_update()
                 .set((
-                    session_questions.eq(1),
-                    (latest_start.eq(now)),
-                    (total_questions.eq(total_questions + 1)),
+                    limits::session_questions.eq(1),
+                    (limits::latest_start.eq(now)),
+                    (limits::total_questions.eq(limits::total_questions + 1)),
                 )),
-            latest_start.le(some_time_ago),
+            limits::latest_start.le(some_time_ago),
         )
-        .returning(session_questions)
+        .returning(limits::session_questions)
         .get_result(conn)
         .optional()?;
         Ok(current)
     }
 }
+
+pub fn add_explain_answer(
+    conn: &mut PgConnection,
+    cache: &AIExplainCacheInsert,
+) -> Result<(), DbError> {
+    insert_into(explain::table)
+        .values(cache)
+        .on_conflict_do_nothing()
+        .execute(conn)?;
+    Ok(())
+}
+
+pub fn explain_from_cache(
+    conn: &mut PgConnection,
+    signature: &Vec<u8>,
+    highlighted_hash: &Vec<u8>,
+) -> Result<Option<AIExplainCacheQuery>, DbError> {
+    let hit = update(explain::table)
+        .filter(
+            explain::signature
+                .eq(signature)
+                .and(explain::highlighted_hash.eq(highlighted_hash))
+                .and(explain::version.eq(AI_EXPLAIN_VERSION)),
+        )
+        .set((
+            explain::last_used.eq(Utc::now().naive_utc()),
+            explain::view_count.eq(explain::view_count + 1),
+        ))
+        .returning(explain::all_columns)
+        .get_result(conn)
+        .optional()?;
+    Ok(hit)
+}
+
+pub fn set_explain_feedback(
+    conn: &mut PgConnection,
+    feedback: ExplainFeedback,
+) -> Result<(), DbError> {
+    let ExplainFeedback {
+        typ,
+        hash,
+        signature,
+    } = feedback;
+    match typ {
+        FeedbackTyp::ThumbsDown => update(explain::table)
+            .filter(
+                explain::signature
+                    .eq(signature)
+                    .and(explain::highlighted_hash.eq(hash))
+                    .and(explain::version.eq(AI_EXPLAIN_VERSION)),
+            )
+            .set(explain::thumbs_down.eq(explain::thumbs_down + 1))
+            .execute(conn)
+            .optional()?,
+        FeedbackTyp::ThumbsUp => update(explain::table)
+            .filter(
+                explain::signature
+                    .eq(signature)
+                    .and(explain::highlighted_hash.eq(hash))
+                    .and(explain::version.eq(AI_EXPLAIN_VERSION)),
+            )
+            .set(explain::thumbs_up.eq(explain::thumbs_up + 1))
+            .execute(conn)
+            .optional()?,
+    };
```

Comment:
Could we do something like this instead? (haven't tested it)

```suggestion
    let op = update(explain::table).filter(
        explain::signature
            .eq(signature)
            .and(explain::highlighted_hash.eq(hash))
            .and(explain::version.eq(AI_EXPLAIN_VERSION)),
    );
    match typ {
        FeedbackTyp::ThumbsDown => op.set(explain::thumbs_down.eq(explain::thumbs_down + 1)),
        FeedbackTyp::ThumbsUp => op.set(explain::thumbs_up.eq(explain::thumbs_up + 1)),
    };
    op.execute(conn).optional()?;
```

Reply:
Similar yes. In diesel these are two different types. This only works when allocating the query on the heap. The way I wrote it's stack allocated, because all types are known at compile time.